// lay/lib/db.js
/**

 @Name：lay db 1.0 for mongodb 操作集合
 @Auther：贤心
 @Date：2014-08-02

*/

var MongoClient = require('mongodb').MongoClient;

var Laydb = function(conn){
    this.dbs = 'mongodb://'+ conn.host + ':' + conn.database[1] + '/' + conn.db;
};

//打开集合
Laydb.prototype.open = function(gather, callback){
    var that = this;
    MongoClient.connect(that.dbs, function(err, db){
        if(err) throw err;
        db.collection(gather, function(err, collect){
            callback(err, collect);
        });
    });
};

//增加集合
Laydb.prototype.insert = function(gather, docs, callback){
    var that = this;
    that.open(gather, function(err, collect){
        collect.insert(docs, function(err, info){
            callback(err, info);
        });
    });
};

//修改集合
Laydb.prototype.update = function(gather, options, callback){
    var that = this;
    that.open(gather, function(err, collect){
        collect.update(options.criteria, options.newdocs, function(err){
            callback(err);
        });
    });
};

//查询单个集合
Laydb.prototype.findOne = function(gather, docs, callback){
    var that = this;
    that.open(gather, function(err, collect){
        collect.findOne(docs, function(err, info){
            callback(err, info);
        });
    });
};

//查询多个集合
Laydb.prototype.find = function(gather, options, callback){
    var that = this;
    var limit = 10;
    that.open(gather, function(err, collect){
        collect.find(options.query)
        .sort(options.sort)
        .limit(options.limit || limit
        ).toArray(function(err, docs) {
            callback(err, docs);
        });
    });
};

module.exports = function(conn){
    return new Laydb(conn);
};
